### ISSUE TITLE
Connect registration APIs to catalogue

### ISSUE THREAD

**Issue by seanprivett** [ministryofjustice/analytical-platform#2779]
registering data product
registering metadata
exclude update

Abstract base class - daughter OMD / DataHub exports

DOD 
Use the python  library and prove end to end working using cmd

**Comment by MatMoore** [comment]
This will involve updating https://github.com/ministryofjustice/data-platform/tree/main/python-libraries/data-platform-catalogue so that we can connect it to DataHub.

Suggestion: keep 2 implementations with the same interface. That way we won't have to delete the OMD one and we can add a DataHub imlementation, while keeping the usage of this library the same, i.e.

```python
  service_fqn = client.create_or_update_database_service(name="data_platform")
  database_fqn = client.create_or_update_database(metadata=data_product, service_fqn=service_fqn)
  schema_fqn = client.create_or_update_schema(metadata=data_product_schema, database_fqn=database_fqn)
  table_fqn = client.create_or_update_table(metadata=table, schema_fqn=schema_fqn)
```

For example, we could turn `CatalogueClient` into a base class and have `OpenMetadataClient` and `DataHubClient` as subclasses.

**Comment by MatMoore** [comment]
Here is some proof of concept code that uses the Datahub API

```python
import datahub.emitter.mce_builder as builder
from datahub.api.entities.dataproduct.dataproduct import DataProduct
from datahub.emitter.mce_builder import make_data_platform_urn, make_dataset_urn
from datahub.emitter.mcp import MetadataChangeProposalWrapper
from datahub.emitter.rest_emitter import DatahubRestEmitter
from datahub.ingestion.graph.client import DatahubClientConfig, DataHubGraph

# This is for interaction with the GraphQL API, which seems to be necessary for newer stuff like Data Products
gms_endpoint = "http://localhost:8080"
graph = DataHubGraph(DatahubClientConfig(server=gms_endpoint))


# Imports for metadata model classes
from datahub.metadata.schema_classes import (
    AuditStampClass,
    DatasetPropertiesClass,
    DateTypeClass,
    NumberTypeClass,
    OtherSchemaClass,
    SchemaFieldClass,
    SchemaFieldDataTypeClass,
    SchemaMetadataClass,
    StringTypeClass,
)

# Create an emitter to DataHub over REST
emitter = DatahubRestEmitter(gms_server="http://localhost:8080", extra_headers={})

# Test the connection
emitter.test_connection()

# Construct a dataset properties object
dataset_properties = DatasetPropertiesClass(
    description="Example table for testing",
    customProperties={"dpia_required": "yes"},
)

# Construct a MetadataChangeProposalWrapper object.
metadata_event = MetadataChangeProposalWrapper(
    entityUrn=builder.make_dataset_urn(
        "athena", "athena.catalog.hmpps_use_of_force_v2.statements"
    ),
    aspect=dataset_properties,
)

emitter.emit(metadata_event)

dataset_schema_properties = SchemaMetadataClass(
    schemaName="customer",  # not used
    platform=make_data_platform_urn("athena"),  # important <- platform must be an urn
    version=1,  # when the source system has a notion of versioning of schemas, insert this in, otherwise leave as 0
    hash="",  # when the source system has a notion of unique schemas identified via hash, include a hash, else leave it as empty string
    # platformSchema: Union["EspressoSchemaClass", "OracleDDLClass", "MySqlDDLClass", "PrestoDDLClass", "KafkaSchemaClass", "BinaryJsonSchemaClass", "OrcSchemaClass", "SchemalessClass", "KeyValueSchemaClass", "OtherSchemaClass"],
    platformSchema=OtherSchemaClass(rawSchema="__insert raw schema here__"),
    lastModified=AuditStampClass(time=1640692800000, actor="urn:li:corpuser:ingestion"),
    fields=[
        SchemaFieldClass(
            fieldPath="statements.id",
            # type: Union["BooleanTypeClass", "FixedTypeClass", "StringTypeClass", "BytesTypeClass", "NumberTypeClass", "DateTypeClass", "TimeTypeClass", "EnumTypeClass", "NullTypeClass", "MapTypeClass", "ArrayTypeClass", "UnionTypeClass", "RecordTypeClass"],
            type=SchemaFieldDataTypeClass(type=NumberTypeClass()),
            nativeDataType="bigint",  # use this to provide the type of the field in the source system's vernacular
            description="Unique identifier for the statement",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.report_id",
            type=SchemaFieldDataTypeClass(type=NumberTypeClass()),
            nativeDataType="bigint",
            description="Unique identifier for the report",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.user_id",
            type=SchemaFieldDataTypeClass(type=NumberTypeClass()),
            nativeDataType="bigint",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.name",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.email",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.submitted_date",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="Date the statement was submitted (unknown format)",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.submitted_date",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.statement_status",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.last_training_month",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.last_training_year",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.job_start_year",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.staff_id",
            type=SchemaFieldDataTypeClass(type=NumberTypeClass()),
            nativeDataType="bigint",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.created_date",
            type=SchemaFieldDataTypeClass(type=DateTypeClass()),
            nativeDataType="timestamp",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.updated_date",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="timestamp",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.next_reminder_date",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.overdue_date",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.in_progress",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.deleted",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.removal_requested_reason",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
        SchemaFieldClass(
            fieldPath="statements.removal_requested_date",
            type=SchemaFieldDataTypeClass(type=StringTypeClass()),
            nativeDataType="string",
            description="???",
            lastModified=AuditStampClass(
                time=1640692800000, actor="urn:li:corpuser:ingestion"
            ),
        ),
    ],
)

metadata_event = MetadataChangeProposalWrapper(
    entityUrn=builder.make_dataset_urn(
        "athena", "athena.catalog.hmpps_use_of_force_v2.statements"
    ),
    aspect=dataset_schema_properties,
)

# Emit metadata! This is a blocking call
emitter.emit(metadata_event)


data_product = DataProduct(
    id="hmpps_use_of_force",
    display_name="Use of force",
    domain="urn:li:domain:4414be1c-8677-4c5b-9799-f955ae9d8432",
    description="TODO - describe the use of force data product 2",
    assets=[
        builder.make_dataset_urn(
            "athena", "athena.catalog.hmpps_use_of_force_v2.statements"
        ),
    ],
    owners=[{"id": "urn:li:corpuser:jdoe", "type": "BUSINESS_OWNER"}],
    terms=["urn:li:glossaryTerm:ClientsAndAccounts.AccountBalance"],
    tags=["urn:li:tag:adoption"],
    properties={"lifecycle": "production", "sla": "7am every day"},
    external_url="https://en.wikipedia.org/wiki/Sloth",
)

for mcp in data_product.generate_mcp(upsert=False):
    graph.emit(mcp)
```

See the docs here: https://datahubproject.io/docs/metadata-ingestion/as-a-library

**Comment by tom-webber** [comment]
In DataHub Data Products require a parent domain. I've implemented logic and methods to create a domain if one does not exists (domain is mandatory metadata for `DataProductMetadata`), and a Data Product if one does not exist when submitting a dataset.

I'd like to implement tests using [the `MockDataHubGraph` class from DataHub's own tests](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/tests/unit/api/entities/dataproducts/test_dataproduct.py#L10C46-L10C62), so I've added the pre-reqs to the PR under `/tests/test_helpers/`.
